File: average-salary-excluding-the-minimum-and-maximum-salary/Solution1491.java
// https://leetcode.com/problems/average-salary-excluding-the-minimum-and-maximum-salary/

class Solution1491 {

    // T: O(n), S: O(1)
    public double average(int[] salary) {
        int n = salary.length;
        int minSal = salary[0], maxSal = salary[0];
        double sum = 0.0;
        for (int sal : salary) {
            minSal = Math.min(minSal, sal);
            maxSal = Math.max(maxSal, sal);
            sum += sal;
        }
        sum -= minSal;
        sum -= maxSal;
        return sum / (n - 2);
    }

    public static void main(String[] args) {
        var sol = new Solution1491();
        var salary1 = new int[] { 4000, 3000, 1000, 2000 };
        var salary2 = new int[] { 1000, 2000, 3000 };
        System.out.println(sol.average(salary1)); // 2500.0
        System.out.println(sol.average(salary2)); // 2000.0
    }

}

File: average-salary-excluding-the-minimum-and-maximum-salary/solution1491.py
# https://leetcode.com/problems/average-salary-excluding-the-minimum-and-maximum-salary/

from typing import List


class Solution1491:

    # T: O(n), S: O(1)
    def average(self, salary: List[int]) -> None:
        min_sal, max_sal, n = salary[0], salary[0], len(salary)
        sum = 0
        for s in salary:
            min_sal = min(min_sal, s)
            max_sal = max(max_sal, s)
            sum += s
        sum -= min_sal
        sum -= max_sal
        return sum / (n - 2)


def main() -> None:
    sol = Solution1491()
    salary1 = [4000, 3000, 1000, 2000]
    salary2 = [1000, 2000, 3000]
    print(sol.average(salary1))  # 2500.0
    print(sol.average(salary2))  # 2000.0


if __name__ == "__main__":
    main()
